### PR TITLE
Background Calculations

### PR DESCRIPTION
**Pull request by hsinfan1996**
Merge distances_v3 branch into master (different from PR #1066 ). Closes issue #954 andd #408 .

### CODE REVIEW

File: pyccl/_core/parameters/parameters_base.py [hunk 1]
```diff
@@ -134,6 +134,10 @@ class GSLParams(CCLParameters, instance=lib.cvar.user_gsl_params):
 class PhysicalConstants(CCLParameters, instance=lib.cvar.constants,
                         freeze=True):
     """Instances of this class hold the physical constants."""
+    # Mean solar day (s/day). [exact]
+    DAY = 86400.
+    # Sidereal year (days/yr). [IERS2014 in J2000.0]
+    YEAR = 365.256_363_004 * DAY
```
Comment (hsinfan1996) [inline pyccl/_core/parameters/parameters_base.py:140]: This part is ad-hoc. We can port all constants from C to Python level.

**Comment by hsinfan1996** [comment]
@nikosarcevic I've done merging the master branch into the revived distances_v3. Now you can try the new functions and see if the calculations are correct and maybe modify the code if you want.

**Comment by damonge** [comment]
@hsinfan1996 let me know when you'd like me to review this. Right now I see the tests are still failing, but happy to help.

**Comment by hsinfan1996** [comment]
@damonge This is ready for review. Is there a way to fix the tests on macOS? I don't know why sometimes they passed but sometimes not.

**Comment by hsinfan1996** [comment]
> @damonge This is ready for review. Is there a way to fix the tests on macOS? I don't know why sometimes they passed but sometimes not.

From my observation, faster test runs (typically < 25 min) on macOS will pass, slower ones (~35 min) will not. The weird thing is that the failed tests are not timed, but it is good to know that the problem is not likely caused by the code.

File: pyccl/background.py [hunk 2]
```diff
@@ -197,6 +226,141 @@ def distance_modulus(cosmo, a):
...
+
+
+def comoving_volume(cosmo, a, *, solid_angle=4*np.pi):
+    r"""Comoving volume, in :math:`\rm Mpc^3`.
+
+    .. math::
+
+        V_{\rm C} = \int_{\Omega} \mathrm{{d}}\Omega \int_z \mathrm{d}z
+        D_{\rm H} \frac{(1+z)^2 D_{\mathrm{A}}^2}{E(z)}
+
```
Comment (damonge) [inline pyccl/background.py:287]:
```suggestion
    See Eq. 29 in `Hogg 2000 <https://arxiv.org/abs/astro-ph/9905116>`_.
    
```

File: pyccl/cosmology.py [hunk 3]
```diff
@@ -101,6 +103,12 @@ def _make_methods(cls=None, *, modules=_TOP_LEVEL_MODULES, name=None):
     return cls
 
 
+@dataclass
+class CosmologyData:
```
Comment (damonge) [inline pyccl/cosmology.py:107]: Can you add a comment here for future developers that the aim is for this class to be used to port other background calculators from C to python?

Also, we should make it a private class (`_CosmologyData`), and perhaps relabel it to `_CosmologyBackgroundData` to make it clearer what it's supposed to hold.

File: pyccl/tests/test_background.py [hunk 4]
```diff
@@ -85,6 +85,29 @@ def test_background_a_interface(a, func):
...
+    # Check that integrating the comoving volume element gives comoving volume.
+    cosmo = ccl.CosmologyVanillaLCDM(Omega_k=Omega_k)
+    from scipy.integrate import quad
+    a_min = 0.4
+    V = cosmo.comoving_volume(a=a_min, solid_angle=1.0)
+    vdv = quad(cosmo.comoving_volume_element, a=a_min, b=1.0)[0]
+    rtol = 1e-7 if Omega_k == 0 else 1e-2
+    assert np.allclose(vdv, V, atol=0, rtol=rtol)
+
+
```
Comment (damonge) [inline pyccl/tests/test_background.py:110]: `comoving_volume_element` isn't being tested anywhere, right?

**Review by damonge — CHANGES_REQUESTED** [review]
Thanks a lot @hsinfan1996 (and @nikfilippas for the initial implementation). Some comments below.

File: pyccl/tests/test_background.py (hunk 4 above)
Comment (damonge) [reply pyccl/tests/test_background.py:110]: OK, yes, good point.

**Review by damonge — APPROVED** [review]
LGTM!

**Pull request merged**